File: src/endstone_primebds/commands/Server/monitor.py
import time
from endstone import Player, ColorFormat
from endstone.command import CommandSender
from endstone_primebds.utils.command_util import create_command
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from endstone_primebds.primebds import PrimeBDS

# Command registration
command, permission = create_command(
    "monitor",
    "Monitor server performance in real time!",
    ["/monitor (server|disable)[debug: debug]"],
    ["primebds.command.monitor"]
)

def handler(self: "PrimeBDS", sender: CommandSender, args: list[str]) -> bool:
    if not isinstance(sender, Player):
        sender.send_error_message("This command can only be executed by a player")
        return True

    player_name = sender.name
    mode = args[0].lower() if args else "server"
    interval = 1.0

    if not hasattr(self, "monitor_intervals"):
        self.monitor_intervals = {}

    if player_name in self.monitor_intervals:
        self.server.scheduler.cancel_task(self.monitor_intervals[player_name])
        del self.monitor_intervals[player_name]

        if not args or args[0].lower() == "disable":
            if len(self.monitor_intervals) == 0:
                self.packets_sent_count.clear()
                self.packet_last_sample = {
                    "time": time.time(),
                    "counts": {}
                }
            sender.send_message(f"{ColorFormat.RED}Monitoring turned off")
            return True

        sender.send_message(f"{ColorFormat.YELLOW}Previous monitoring canceled, applying new settings...")

    def monitor_interval(player_name, mode=mode):
        player = self.server.get_player(player_name)
        if not player:
            if player_name in self.monitor_intervals:
                self.server.scheduler.cancel_task(self.monitor_intervals[player_name])
                del self.monitor_intervals[player_name]
            return

        if mode == "server":
            overworld = self.server.level.get_dimension("Overworld")
            nether = self.server.level.get_dimension("Nether")
            the_end = self.server.level.get_dimension("TheEnd")

            tps = self.server.average_tps
            mspt = self.server.average_mspt
            mspt_cur = self.server.current_mspt
            tick_usage = self.server.average_tick_usage
            entity_count = len(self.server.level.actors)
            server_version = self.server.minecraft_version
            overworld_chunks = len(overworld.loaded_chunks) if overworld else 0
            nether_chunks = len(nether.loaded_chunks) if nether else 0
            the_end_chunks = len(the_end.loaded_chunks) if the_end else 0

            tps_display = int(tps)
            tps_fraction = int((tps - tps_display) * 10)
            tps_color = ColorFormat.GREEN if tps > 18 else ColorFormat.YELLOW if 14 <= tps <= 18 else ColorFormat.RED
            ping_color = get_ping_color(player.ping)
            mspt_color = ColorFormat.GREEN if mspt < 50 else ColorFormat.RED
            mspt_cur_color = ColorFormat.GREEN if mspt_cur < 50 else ColorFormat.RED
            entity_color = ColorFormat.GREEN if entity_count < 600 else ColorFormat.YELLOW if entity_count <= 800 else ColorFormat.RED
            dim_color = {"Overworld": ColorFormat.GREEN, "Nether": ColorFormat.RED, "TheEnd": ColorFormat.MATERIAL_IRON}.get(player.dimension.name, ColorFormat.GRAY)

            tps_str = f"{tps_color}{tps_display}.{tps_fraction:1d} {ColorFormat.ITALIC}{ColorFormat.GRAY}({tick_usage:.1f})"
            ping_str = f"{ping_color}{player.ping}ms"
            mspt_str = f"{mspt_color}{mspt:.1f}ms {ColorFormat.ITALIC}{ColorFormat.GRAY}(avg) {ColorFormat.RESET}{ColorFormat.GRAY}| {mspt_cur_color}{mspt_cur:.1f}ms {ColorFormat.ITALIC}{ColorFormat.GRAY}(cur)"
            entity_str = f"{entity_color}{entity_count}"
            version_str = f"{ColorFormat.GREEN}{server_version}"
            chunk_str = f"{ColorFormat.GREEN}{overworld_chunks} {ColorFormat.GRAY}| {ColorFormat.RED}{nether_chunks} {ColorFormat.GRAY}| {ColorFormat.MATERIAL_IRON}{the_end_chunks}"
            your_dim = f"{dim_color}{player.dimension.name}"

            player.send_tip(
                f"{ColorFormat.AQUA}Server Monitor{ColorFormat.RESET}\n"
                f"{ColorFormat.RESET}---------------------------\n"
                f"{ColorFormat.RESET}Level: {self.server.level.name} {ColorFormat.ITALIC}{ColorFormat.GRAY}(ver. {version_str}{ColorFormat.GRAY})\n"
                f"{ColorFormat.RESET}TPS: {tps_str} {ColorFormat.RESET}\n"
                f"{ColorFormat.RESET}MSPT: {mspt_str}\n"
                f"{ColorFormat.RESET}Loaded Chunks: {chunk_str}\n"
                f"{ColorFormat.RESET}Loaded Entities: {entity_str}\n"
                f"{ColorFormat.RESET}---------------------------\n"
                f"{ColorFormat.RESET}Your Ping: {ping_str}\n"
                f"{ColorFormat.RESET}Current DIM: {your_dim}"
            )

    task = self.server.scheduler.run_task(
        self,
        lambda: monitor_interval(player_name, mode),
        delay=0,
        period=int(interval * 20)
    )

    if task:
        self.monitor_intervals[player_name] = task.task_id
        sender.send_message(f"{ColorFormat.GREEN}Monitoring turned on with {interval:.1f}s interval ({mode} mode)")
    else:
        sender.send_error_message("Failed to start monitoring task.")

    return True

def get_ping_color(ping: int) -> str:
    return ColorFormat.GREEN if ping <= 80 else ColorFormat.YELLOW if ping <= 160 else ColorFormat.RED

def clear_all_monitor_intervals(self: "PrimeBDS"):
    """Clear all active intervals."""
    for player_name, task_id in self.monitor_intervals.items():
        self.server.scheduler.cancel_task(task_id) 
    self.monitor_intervals.clear()

File: src/commands/server/monitor.cpp
#include "primebds/commands/command_registry.h"
#include "primebds/plugin.h"

#include <cstdio>

namespace primebds::commands
{

    static const char *get_tps_color(float tps)
    {
        if (tps > 18.0f)
            return "\u00a7a"; // green
        if (tps >= 14.0f)
            return "\u00a7e"; // yellow
        return "\u00a7c";     // red
    }

    static const char *get_ping_color(int ping)
    {
        if (ping <= 80)
            return "\u00a7a"; // green
        if (ping <= 160)
            return "\u00a7e"; // yellow
        return "\u00a7c";     // red
    }

    static const char *get_mspt_color(float mspt)
    {
        return mspt < 50.0f ? "\u00a7a" : "\u00a7c";
    }

    static const char *get_entity_color(int count)
    {
        if (count < 600)
            return "\u00a7a"; // green
        if (count <= 800)
            return "\u00a7e"; // yellow
        return "\u00a7c";     // red
    }

    static const char *get_dim_color(const std::string &dim_name)
    {
        if (dim_name == "Overworld")
            return "\u00a7a";
        if (dim_name == "Nether")
            return "\u00a7c";
        if (dim_name == "TheEnd")
            return "\u00a77";
        return "\u00a78";
    }

    static void monitor_tick(PrimeBDS &plugin, const std::string &player_name)
    {
        auto *player = plugin.getServer().getPlayer(player_name);
        if (!player)
        {
            plugin.monitor_intervals.erase(player_name);
            return;
        }

        auto &server = plugin.getServer();
        auto *level = server.getLevel();
        if (!level)
            return;

        float tps = server.getAverageTicksPerSecond();
        float mspt_avg = server.getAverageMillisecondsPerTick();
        float mspt_cur = server.getCurrentMillisecondsPerTick();
        float tick_usage = server.getAverageTickUsage();
        int entity_count = static_cast<int>(level->getActors().size());
        std::string server_version = server.getMinecraftVersion();
        std::string level_name = level->getName();

        // Chunk counts per dimension
        auto *overworld = level->getDimension("Overworld");
        auto *nether = level->getDimension("Nether");
        auto *the_end = level->getDimension("TheEnd");
        int overworld_chunks = overworld ? static_cast<int>(overworld->getLoadedChunks().size()) : 0;
        int nether_chunks = nether ? static_cast<int>(nether->getLoadedChunks().size()) : 0;
        int end_chunks = the_end ? static_cast<int>(the_end->getLoadedChunks().size()) : 0;

        // Player stats
        int ping = static_cast<int>(player->getPing().count());
        std::string player_dim = player->getDimension().getName();

        int tps_int = static_cast<int>(tps);
        int tps_frac = static_cast<int>((tps - tps_int) * 10);

        char buf[1024];
        std::snprintf(buf, sizeof(buf),
                      "\u00a7bServer Monitor\u00a7r\n"
                      "\u00a7r---------------------------\n"
                      "\u00a7rLevel: %s \u00a7o\u00a77(ver. \u00a7a%s\u00a77)\n"
                      "\u00a7rTPS: %s%d.%d \u00a7o\u00a77(%.1f%%)\u00a7r\n"
                      "\u00a7rMSPT: %s%.1fms \u00a7o\u00a77(avg) \u00a7r\u00a77| %s%.1fms \u00a7o\u00a77(cur)\n"
                      "\u00a7rLoaded Chunks: \u00a7a%d \u00a77| \u00a7c%d \u00a77| \u00a77%d\n"
                      "\u00a7rLoaded Entities: %s%d\n"
                      "\u00a7r---------------------------\n"
                      "\u00a7rYour Ping: %s%dms\n"
                      "\u00a7rCurrent DIM: %s%s",
                      level_name.c_str(),
                      server_version.c_str(),
                      get_tps_color(tps), tps_int, tps_frac, tick_usage,
                      get_mspt_color(mspt_avg), mspt_avg, get_mspt_color(mspt_cur), mspt_cur,
                      overworld_chunks, nether_chunks, end_chunks,
                      get_entity_color(entity_count), entity_count,
                      get_ping_color(ping), ping,
                      get_dim_color(player_dim), player_dim.c_str());

        player->sendTip(buf);
    }

    static bool cmd_monitor(PrimeBDS &plugin, endstone::CommandSender &sender,
                            const std::vector<std::string> &args)
    {
        auto *player = sender.asPlayer();
        if (!player)
        {
            sender.sendMessage("\u00a7cOnly players can use this command.");
            return true;
        }

        std::string player_name = player->getName();
        bool is_active = plugin.monitor_intervals.count(player_name) > 0;

        // Resolve desired state: explicit arg overrides, otherwise toggle
        bool want_on;
        if (!args.empty())
        {
            std::string arg = args[0];
            for (auto &c : arg)
                c = static_cast<char>(std::tolower(c));

            if (arg == "true" || arg == "on" || arg == "1" || arg == "enable" || arg == "server")
                want_on = true;
            else if (arg == "false" || arg == "off" || arg == "0" || arg == "disable")
                want_on = false;
            else
                want_on = !is_active;
        }
        else
        {
            want_on = !is_active;
        }

        // Disable
        if (!want_on)
        {
            if (is_active)
            {
                plugin.getServer().getScheduler().cancelTask(plugin.monitor_intervals[player_name]);
                plugin.monitor_intervals.erase(player_name);
                sender.sendMessage("\u00a7cMonitoring turned off");
            }
            else
            {
                sender.sendMessage("\u00a7cMonitoring is not active");
            }
            return true;
        }

        // If already active, cancel first then restart
        if (is_active)
        {
            plugin.getServer().getScheduler().cancelTask(plugin.monitor_intervals[player_name]);
            plugin.monitor_intervals.erase(player_name);
        }

        // Schedule repeating task (every 20 ticks = 1 second)
        auto task = plugin.getServer().getScheduler().runTaskTimer(
            plugin,
            [&plugin, player_name]()
            {
                monitor_tick(plugin, player_name);
            },
            0, 5);

        if (task)
        {
            plugin.monitor_intervals[player_name] = task->getTaskId();
            sender.sendMessage("\u00a7aMonitoring turned on (server mode)");
        }
        else
        {
            sender.sendMessage("\u00a7cFailed to start monitoring task.");
        }

        return true;
    }

    REGISTER_COMMAND(monitor, "Monitor server performance in real time!", cmd_monitor,
                     info.usages = {"/monitor [enable: bool]"};
                     info.permissions = {"primebds.command.monitor"};);
} // namespace primebds::commands
